### PR TITLE
o/snapstate: migrate on core22 refresh and init ~/Snap

### PR DESCRIPTION
**Pull request by MiguelPires**
This PR:
* hides/migrates snaps that are refreshing to a revision based on core22 (or later) for the first time
* creates and initializes ~/Snap based on the last state of SNAP_USER_DATA
* sets env vars based on the new directories

 Opening this as draft PR to get feedback from @pedronis. Once ready, I'll split this into several PRs.

### CODE REVIEW

**Comment by MiguelPires** [comment]
Just a few notes on things missing from this PR.
This doesn't deal with reverting the ~/Snap creation and initialization. Since ~/Snap is split off from SNAP_USER_DATA, it's unclear what the revert behavior should be. Deleting it might discard user data but merging also introduces its own problems.
Existing files in SNAP_USER_DATA can clash with the new XDG dirs (e.g., a file named ~/snap/<name>/x1/config would conflict with the XDG_CONFIG_HOME dir creation during the migration). One option to deal with this would be to not add the XDG_*_HOME dir and env var, if the path exists and is not a directory (or a link to a directory). Also, if the snap already maps an XDG_*_HOME env var in its yaml, do we still want to create the corresponding directory? I'd assume so, for consistency, but one might argue that the mapping means it won't be used so the dir is unnecessary.

File: overlord/snapstate/handlers.go [hunk 1]
```diff
@@ -1205,6 +1214,35 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
+func shouldHideSnapData(opts *dirMigrationOptions, newBase string) bool {
+	// hide snap data if any is true:
+	//	* 1st refresh after user set experimental flag
+	//	* 1st refresh to revision w/ core22 base (or newer)
+	return !opts.MigratedToHidden && (opts.UseHidden || atLeastCore22(newBase))
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:1221]: this condition looks right

File: cmd/snap/cmd_run.go [hunk 2]
```diff
@@ -403,6 +403,14 @@ func createUserDataDirs(info *snap.Info, opts *dirs.SnapDirOptions) error {
 	instanceUserData := info.UserDataDir(usr.HomeDir, opts)
 	instanceCommonUserData := info.UserCommonDataDir(usr.HomeDir, opts)
 	createDirs := []string{instanceUserData, instanceCommonUserData}
+
+	if opts.HiddenSnapDataDir {
```
Comment (pedronis) [inline cmd/snap/cmd_run.go:407]: we shouldn't do this if the base is not core22+, no?

File: snap/snapenv/snapenv.go [hunk 3]
```diff
@@ -122,5 +126,14 @@ func userEnv(info *snap.Info, home string, opts *dirs.SnapDirOptions) osutil.Env
 	}
 	// Provide the location of the real home directory.
 	env["SNAP_REAL_HOME"] = home
+
+	if opts.HiddenSnapDataDir {
```
Comment (pedronis) [inline snap/snapenv/snapenv.go:130]: same, we shouldn't do this if the base is not core22+, no?

File: overlord/snapstate/handlers.go [hunk 4]
```diff
@@ -1177,7 +1182,11 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
 		if err != nil {
 			return fmt.Errorf("cannot set migration status to done: %w", err)
 		}
-	} else if !opts.UseHidden && opts.MigratedToHidden {
+
+		if err := m.backend.InitSnapUserHome(snapsup.InstanceName(), newInfo.Revision); err != nil {
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:1186]: again, I'm not sure I understand, we shouldn't do this for non-core22+ snaps?

**Review by pedronis — COMMENTED** [review]
I'm confused, it seems we are using ~/Snap even for snaps that don't have core22+ base, we want to be able to hide any snaps, but only use hide and use ~/Snap for core22+ snaps

File: cmd/snap/cmd_run.go (hunk 2 above)
Comment (pedronis) [reply cmd/snap/cmd_run.go:407]: I would have expected a new and seperate flag UseExposedUserSnapDir in place or something, that can be set only for snaps with core22+ bases, though it would always be the case that if that is true then HiddenSnapDataDir is true as well

**Comment by MiguelPires** [comment]
> I'm confused, it seems we are using ~/Snap even for snaps that don't have core22+ base, we want to be able to hide any snaps, but only use hide and use ~/Snap for core22+ snaps

My understanding was that the experimental migration path was going to have the same behavior as the core22 migration path.  So the experimental stuff would evolve with the core22 to serve as an early test for it. Anyway, if not, then I'll just have to shuffle some bits around and re-do some tests. I'll do that and ping you ASAP

File: snap/info.go [hunk 5]
```diff
@@ -526,11 +530,17 @@ func (s *Info) UserCommonDataDir(home string, opts *dirs.SnapDirOptions) string
 	return UserCommonDataDir(home, s.InstanceName(), opts)
 }
 
+// NewSnapUserHome returns the new upper-case snap directory in the user home.
+func (s *Info) NewSnapUserHome(home string) string {
```
Comment (pedronis) [inline snap/info.go:534]: we should find some consistent naming for this, something like:

UserExposedHomeDir
UserExposedDataDir

File: snap/info.go [hunk 6]
```diff
@@ -228,6 +228,10 @@ func UserSnapDir(home string, name string, opts *dirs.SnapDirOptions) string {
 	return filepath.Join(home, snapDataDir(opts), name)
 }
 
+func ExposedUserSnapDir(home string, name string) string {
```
Comment (pedronis) [inline snap/info.go:231]: this and other places should then be consistent

File: dirs/dirs.go [hunk 7]
```diff
@@ -167,6 +171,8 @@ var (
 type SnapDirOptions struct {
 	// HiddenSnapDataDir determines if the snaps' data is in ~/.snap/data instead of ~/snap
 	HiddenSnapDataDir bool
+	// UseExposedDir determines if ~/Snap is in use.
+	UseExposedDir bool
```
Comment (pedronis) [inline dirs/dirs.go:175]: this should have a name that matches a bit more the ones we pick in the snap package

File: overlord/snapstate/handlers.go [hunk 8]
```diff
@@ -1150,9 +1150,43 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
...
-	// the migration hasn't been done - do it now
-	if opts.UseHidden && !opts.MigratedToHidden {
-		if err = m.backend.HideSnapData(snapsup.InstanceName()); err != nil {
+	var oldBase string
+	if oldInfo != nil {
+		oldBase = oldInfo.Base
+	}
+
+	// snap is already migrated, the flag has been turned off but we're
+	// refreshing to a core22 based revision for the 1st time
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:1159]: this is mixing predicates helpers and predicates in the ifs,  maybe it would be better to have a single function that returns a couple of booleans? or return an enumeration of possible states/actions and then there's  a switch here?

just a suggestion but this is a bit hard to follow

**Review by pedronis — COMMENTED** [review]
thanks for the changes, this seems to be going in the right/agreed direction now. I made some high level naming and code structure comments. I think we should think how to split this up at least in two or 3 PRs for easier review.

**Comment by MiguelPires** [comment]
I'm closing this since the PRs split off from this draft have been merged
